Fixed #35103 -- Used provided error code and message when fields is set without a condition on UniqueConstraint. (#17723)
django/django: docs/ref/models/constraints.txt

.. attribute:: UniqueConstraint.violation_error_code

The error code used when a ``ValidationError`` is raised during
:ref:`model validation <validating-objects>`.

> If :attr:`~UniqueConstraint.fields` is set and without a

Reviewer: If you like my other suggestion, then this would be
```
Defaults to :attr:`.BaseConstraint.violation_error_code`, when either
:attr:`.UniqueConstraint.condition` is provided or
:attr:`.UniqueConstraint.fields` is not provided.

When :attr:`.UniqueConstraint.fields` is provided without a
:attr:`.UniqueConstraint.condition`, then if :attr:`.UniqueConstraint.fields`
references multiple fields, defaults to the
:attr:`Meta.unique_together <django.db.models.Options.unique_together>` error
code. Otherwise defaults to the :attr:`.Field.unique` error code.
```

:attr:`~UniqueConstraint.condition`, the
:attr:`~django.db.models.UniqueConstraint.violation_error_code` defaults to the
same error code as when setting :attr:`.Field.unique` when there is a single
field, and to the same error code as when setting
:attr:`Meta.unique_together <django.db.models.Options.unique_together>` when
there are multiple.

Otherwise, defaults to :attr:`.BaseConstraint.violation_error_message`.

.. versionchanged:: 5.2

> If :attr:`~django.db.models.UniqueConstraint.violation_error_code` is
> set, this code is now always used.

Reviewer: ```suggestion
    In older versions, the provided
    :attr:`.UniqueConstraint.violation_error_code` was not used when
    :attr:`.UniqueConstraint.fields` was set.
```
We need an indent and write this in past tense

``violation_error_message``
---------------------------

.. attribute:: UniqueConstraint.violation_error_message

The error message used when a ``ValidationError`` is raised during
:ref:`model validation <validating-objects>`.

> If :attr:`~UniqueConstraint.fields` is set and without a
> :attr:`~UniqueConstraint.condition`, the
> :attr:`~django.db.models.UniqueConstraint.violation_error_message` defaults to
> the same error message as when setting :attr:`.Field.unique` when there is a
> single field, and to the same error message as when setting
> :attr:`Meta.unique_together <django.db.models.Options.unique_together>` when
> there are multiple.
> 
> Otherwise, defaults to :attr:`.BaseConstraint.violation_error_message`.

Reviewer: This is still quite a mouth-full. Do you think this is clearer?
```suggestion
Defaults to :attr:`.BaseConstraint.violation_error_message`, when either
:attr:`.UniqueConstraint.condition` is provided or
:attr:`.UniqueConstraint.fields` is not provided.

When :attr:`.UniqueConstraint.fields` is provided without a
:attr:`.UniqueConstraint.condition`, then if :attr:`.UniqueConstraint.fields`
references multiple fields, defaults to the
:attr:`Meta.unique_together <django.db.models.Options.unique_together>` error
message. Otherwise defaults to the :attr:`.Field.unique` error message.
```

.. versionchanged:: 5.2

> If :attr:`~django.db.models.UniqueConstraint.violation_error_message` is
> set, this message is now always used.

Reviewer: ```suggestion
    In older versions, the provided
    :attr:`.UniqueConstraint.violation_error_message` was not used when
    :attr:`.UniqueConstraint.fields` was set.
```

.. attribute:: UniqueConstraint.violation_error_code

> The error code used when ``ValidationError`` is raised during
> :ref:`model validation <validating-objects>`. The error code of such :class:`~UniqueConstraint`\s 
> will then be different from :attr:`.Field.unique` or in
> :attr:`Meta.unique_together <django.db.models.Options.unique_together>` if a custom 
> violation_error_code is set. Defaults to
> :attr:`.BaseConstraint.violation_error_message`.

Reviewer: ```suggestion
The error code used when a ``ValidationError`` is raised during
:ref:`model validation <validating-objects>`.

If :attr:`~UniqueConstraint.fields` is set and without a
:attr:`~UniqueConstraint.condition`, the
:attr:`~django.db.models.UniqueConstraint.violation_error_code` defaults to the
same error code as when setting :attr:`.Field.unique` when there is a single
field, and to the same error code as when setting
:attr:`Meta.unique_together <django.db.models.Options.unique_together>` when
there are multiple.

Otherwise, defaults to :attr:`.BaseConstraint.violation_error_message`.
```

> :attr:`Meta.unique_together <django.db.models.Options.unique_together>` if a custom 
> violation_error_code is set. Defaults to
> :attr:`.BaseConstraint.violation_error_message`.
> 
> .. versionchanged:: 5.2
> 
> This code *is* also used for :class:`UniqueConstraint`\s with
> :attr:`~UniqueConstraint.fields` and without a
> :attr:`~UniqueConstraint.condition`. The error code of such :class:`~UniqueConstraint`\s 
> will then be different from :attr:`.Field.unique` or in
> :attr:`Meta.unique_together <django.db.models.Options.unique_together>` if a custom 
> violation_error_code is set.

Reviewer: ```suggestion
    If :attr:`~django.db.models.UniqueConstraint.violation_error_code` is
    set, this code is now always used.
```

``violation_error_message``
---------------------------

.. attribute:: UniqueConstraint.violation_error_message

> The error message used when ``ValidationError`` is raised during
> :ref:`model validation <validating-objects>`. The error message of such 
> :class:`~UniqueConstraint`\s will then be different from constraints defined 
> with :attr:`.Field.unique` or in :attr:`Meta.unique_together 
> <django.db.models.Options.unique_together>`. Defaults to
> :attr:`.BaseConstraint.violation_error_message`.

Reviewer: ```suggestion
The error message used when a ``ValidationError`` is raised during
:ref:`model validation <validating-objects>`.

If :attr:`~UniqueConstraint.fields` is set and without a
:attr:`~UniqueConstraint.condition`, the
:attr:`~django.db.models.UniqueConstraint.violation_error_message` defaults to
the same error message as when setting :attr:`.Field.unique` when there is a
single field, and to the same error message as when setting
:attr:`Meta.unique_together <django.db.models.Options.unique_together>` when
there are multiple.

Otherwise, defaults to :attr:`.BaseConstraint.violation_error_message`.
```

.. versionchanged:: 5.2

> This message *is* also used for :class:`UniqueConstraint`\s with
> :attr:`~UniqueConstraint.fields` and without a
> :attr:`~UniqueConstraint.condition`. The error message of such 
> :class:`~UniqueConstraint`\s will then be different from constraints defined 
> with :attr:`.Field.unique` or in :attr:`Meta.unique_together 
> <django.db.models.Options.unique_together>`.

Reviewer: ```suggestion
    If :attr:`~django.db.models.UniqueConstraint.violation_error_message` is
    set, this message is now always used.
```

The error code used when ``ValidationError`` is raised during

> :ref:`model validation <validating-objects>`. Defaults to ``None``.

Reviewer: I think we can make these docs updates clearer 🙂 

Firstly, I think we should have a `versionchanged` rather than `versionadded` note and be aware that if someone doesn't read the note, they shouldn't lose any information on how this works.

We have information of the default behaviour saying here `Defaults to None` and below `Defaults to
:attr:.BaseConstraint.violation_error_message.`. This is not true in the case that `:attr:~UniqueConstraint.fields` is used without a `:attr:~UniqueConstraint.condition` and I think this should be spelled out that there are 2 default cases. 
If `unique_together` does get deprecated (as the docs mention it might in future), then this default behaviour would get simplified but currently there are two default messages.

Reviewer: > Firstly, I think we should have a `versionchanged` rather than `versionadded` note and be aware that if someone doesn't read the note, they shouldn't lose any information on how this works.

This is done thank you 👍 

> We have information of the default behaviour saying here `Defaults to None` and below `Defaults to :attr:.BaseConstraint.violation_error_message.`. This is not true in the case that `:attr:~UniqueConstraint.fields` is used without a `:attr:~UniqueConstraint.condition` and I think this should be spelled out that there are 2 default cases. If `unique_together` does get deprecated (as the docs mention it might in future), then this default behaviour would get simplified but currently there are two default messages.

This is not done, let me try to explain.
```
:ref:`model validation <validating-objects>`. Defaults to ``None``.
```
There is no reference for the case this is **not true** and explaining when the error code is the same as when using `Field.unique` or `Meta.unique_together`.

We shouldn't have information explaining how it works contained in the versionchanged note that doesn't exist in the definition (as these notes get deleted in time).

Same comment for `.. attribute:: UniqueConstraint.violation_error_message` 👍

:ref:`model validation <validating-objects>`. Defaults to
:attr:`.BaseConstraint.violation_error_message`.

.. versionadded:: 5.1

This message *is* also used for :class:`UniqueConstraint`\s with
:attr:`~UniqueConstraint.fields` and without a
:attr:`~UniqueConstraint.condition`. The error message of such 
:class:`~UniqueConstraint`\s will then be different from constraints defined 
with :attr:`.Field.unique` or in :attr:`Meta.unique_together 

> <django.db.models.Options.constraints>`.

Reviewer: To me it looks like `unique_together` is linked to the wrong place, I think it should link to [`unique_together`](https://docs.djangoproject.com/en/5.0/ref/models/options/#unique-together)

Reviewer: > To me it looks like unique_together is linked to the wrong place, I think it should link to [unique_together](https://docs.djangoproject.com/en/5.0/ref/models/options/#unique-together)

This is not addressed and this would also be the case above
